test/Atmos/EDMF/edmf_kernels.jl

```diff
@@ -761,28 +797,34 @@ function flux(::Advect{up_ρa{i}}, atmos, args) where {i}
     @unpack state, aux = args
     up = state.turbconv.updraft
     ẑ = vertical_unit_vector(atmos, aux)
-    return up[i].ρaw * ẑ
+    return ρa_up[i] < eps(FT) ? FT(0) * ẑ : up[i].ρaw * ẑ
 end
 function flux(::Advect{up_ρaw{i}}, atmos, args) where {i}
     @unpack state, aux = args
     @unpack ρa_up = args.precomputed.turbconv
+    FT = eltype(state)
     up = state.turbconv.updraft
     ẑ = vertical_unit_vector(atmos, aux)
-    return up[i].ρaw * up[i].ρaw / ρa_up[i] * ẑ
+    return ρa_up[i] < eps(FT) ? FT(0) * ẑ :
+           up[i].ρaw * up[i].ρaw / ρa_up[i] * ẑ
 end
 function flux(::Advect{up_ρaθ_liq{i}}, atmos, args) where {i}
     @unpack state, aux = args
     @unpack ρa_up = args.precomputed.turbconv
+    FT = eltype(state)
     up = state.turbconv.updraft
     ẑ = vertical_unit_vector(atmos, aux)
-    return up[i].ρaw / ρa_up[i] * up[i].ρaθ_liq * ẑ
+    return ρa_up[i] < eps(FT) ? FT(0) * ẑ :
+           up[i].ρaw / ρa_up[i] * up[i].ρaθ_liq * ẑ
 end
 function flux(::Advect{up_ρaq_tot{i}}, atmos, args) where {i}
     @unpack state, aux = args
     @unpack ρa_up = args.precomputed.turbconv
+    FT = eltype(state)
```

Comment:
```suggestion
    FT = eltype(atmos)
```
IIRC, we somehow lose type-inference of for this on the GPU. We should probably look into it, but let's just grab `FT` from `atmos` for now, since that's not packed into and passed via a `NamedTuple`.

Reply:
What about functions where `atmos `is not passed, such as `precompute` ?